book/scheduling-and-threading.md

```diff
@@ -1336,93 +1298,73 @@ class Browser:
+thread-safe, but SDL may not be.
+
+[^why-locks]: Despite the global interpreter lock, we still need
+locks. Each Python thread can still yield between bytecode operations,
+so you can still get concurrent accesses to shared variables, and race
+conditions are still possible. And in fact, while debugging the code
+for this chapter, I encountered this kind of race condition when I
+forgot to add a lock; try removing some of the locks from your browser
+to see for yourself!
 
-[gil]: https://wiki.python.org/moin/GlobalInterpreterLock
 
 Threaded scrolling
 ==================
 
-Two sections back we talked about how important it is to run scrolling on the 
-browser thread, to avoid slow scripts. But right now, even though there
-is a browser thread, scrolling still happens in a `Task` on the main thread.
-Let's now make scrolling truly *threaded*, meaning it runs on the browser
-thread in parallel with scripts and other main thread work.
-
-Threaded scrolling is quite tricky to implement. The reason is that both the
-browser thread *and* the main thread can affect scroll. For example, when
-loading a new page, scroll is set to 0; when running `innerHTML`, the height of
-the document could change, leading to a potential change of scroll offset to
-clamp it to that height. What should we do if the two threads disagree about
-the scroll offset?
-
-The best policy is to respect the scroll offset the user last observed, unless
-it's incompatible with the web page. In other words, use the browser thread
-scroll, unless a new web page has loaded or the scroll exceeds the current
-document height. Let's implement that.
-
-The trickiest part is how to communicate a browser thread scroll offset to the
-main thread, and integrate it with the rendering pipeline. It'll work like
-this:
+Splitting the main thread from the browser thread means that the main
+thread can run a lot of JavaScript without slowing down the browser
+much. But it's still possible for really slow JavaScript to slow the
+browser down. For example, imagine our counter adds the following
+artificial slowdown:
 
-* When the browser thread changes scroll offset, store it in a `scroll` variable
-  on the `Browser`, set `needs_raster_and_draw`, and schedule an animation
-  frame. This will immediately apply the scroll offset to the screen (at the
-  next time `raster_and_draw` is called from the browser thread event loop).
-  Scheduling an animation frame is necessary only to notify the main thread
-  that scroll was changed.
+``` {.javascript file=eventloop}
+function callback() {
+    for (var i = 0; i < 1e9; i++);
+    // ...
+}
+```
+
+Now, every tick of the counter has an artificial pause during which
+the main thread is stuck running JavaScript. This means it can't
+respond to any events; for example, if you hold down the down array,
+the scrolling will be janky and annoying. To fix this, we need to move
+scrolling from the main thread to the browser thread.
```

Comment:
"I encourage you to try it for yourself and witness the annoyance."

This point is important IMO.